File: UnityProject/Assets/Simulation/Core/GoalNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;

namespace HistoryTeller.Simulation
{
    /// <summary>Дерево условий цели уровня: all / any / not / flag / relation.</summary>
    public abstract class GoalNode
    {
        public abstract bool IsMet(World world);

        public static GoalNode Parse(JObject obj)
        {
            if (obj["all"] != null)
                return new AllGoal(obj["all"].Select(t => Parse((JObject)t)).ToList());
            if (obj["any"] != null)
                return new AnyGoal(obj["any"].Select(t => Parse((JObject)t)).ToList());
            if (obj["not"] != null)
                return new NotGoal(Parse((JObject)obj["not"]));
            if (obj["flag"] != null)
            {
                var f = obj["flag"];
                return new FlagGoal((string)f["char"], (string)f["is"]);
            }
            if (obj["relation"] != null)
            {
                var r = obj["relation"];
                return new RelationGoal((string)r["rel"], (string)r["from"], (string)r["to"]);
            }
            throw new ArgumentException("Unknown goal node: " + obj);
        }
    }

    public sealed class AllGoal : GoalNode
    {
        public readonly List<GoalNode> Children;
        public AllGoal(List<GoalNode> children) { Children = children; }
        public override bool IsMet(World w) => Children.All(c => c.IsMet(w));
    }

    public sealed class AnyGoal : GoalNode
    {
        public readonly List<GoalNode> Children;
        public AnyGoal(List<GoalNode> children) { Children = children; }
        public override bool IsMet(World w) => Children.Any(c => c.IsMet(w));
    }

    public sealed class NotGoal : GoalNode
    {
        public readonly GoalNode Child;
        public NotGoal(GoalNode child) { Child = child; }
        public override bool IsMet(World w) => !Child.IsMet(w);
    }

    public sealed class FlagGoal : GoalNode
    {
        public readonly string Char;
        public readonly string Flag;
        public FlagGoal(string charId, string flag) { Char = charId; Flag = flag; }
        public override bool IsMet(World w) => w.HasFlag(Char, Flag);
    }

    public sealed class RelationGoal : GoalNode
    {
        public readonly string Rel;
        public readonly string From;
        public readonly string To;
        public RelationGoal(string rel, string from, string to) { Rel = rel; From = from; To = to; }
        public override bool IsMet(World w) => w.HasRelation(Rel, From, To);
    }
}

File: ios/Modules/Simulation/Sources/Simulation/GoalNode.swift
import Foundation

/// Дерево условий цели уровня: all / any / not / flag / relation.
public indirect enum GoalNode {
    case all([GoalNode])
    case any([GoalNode])
    case not(GoalNode)
    case flag(char: String, flag: String)
    case relation(rel: String, from: String, to: String)

    public func isMet(_ world: World) -> Bool {
        switch self {
        case .all(let children):  return children.allSatisfy { $0.isMet(world) }
        case .any(let children):  return children.contains { $0.isMet(world) }
        case .not(let child):     return !child.isMet(world)
        case .flag(let char, let flag):
            return world.hasFlag(char, flag)
        case .relation(let rel, let from, let to):
            return world.hasRelation(rel, from, to)
        }
    }
}

extension GoalNode: Decodable {
    private enum CodingKeys: String, CodingKey { case all, any, not, flag, relation }
    private struct FlagObj: Decodable { let char: String; let `is`: String }
    private struct RelObj: Decodable { let rel: String; let from: String; let to: String }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if c.contains(.all) {
            self = .all(try c.decode([GoalNode].self, forKey: .all))
        } else if c.contains(.any) {
            self = .any(try c.decode([GoalNode].self, forKey: .any))
        } else if c.contains(.not) {
            self = .not(try c.decode(GoalNode.self, forKey: .not))
        } else if c.contains(.flag) {
            let f = try c.decode(FlagObj.self, forKey: .flag)
            self = .flag(char: f.char, flag: f.is)
        } else if c.contains(.relation) {
            let r = try c.decode(RelObj.self, forKey: .relation)
            self = .relation(rel: r.rel, from: r.from, to: r.to)
        } else {
            throw DecodingError.dataCorrupted(.init(
                codingPath: decoder.codingPath,
                debugDescription: "Unknown goal node"))
        }
    }
}
